hwy/tests/hwy_gtest.h

```diff
@@ -174,18 +178,366 @@ std::string TestParamTargetNameAndT(
+             '\0' &&
+         filter_pattern_end_ch != ':' &&
+         (filter_pattern_end_ch != '-' ||
+          ptr_to_vector_to_append_pattern_to ==
+              ptr_to_negative_test_filter_patterns);
+         ++filter_pattern_len) {
+    }
+
+    // Add the current filter pattern to *ptr_to_vector_to_append_pattern_to
+    ptr_to_vector_to_append_pattern_to->emplace_back(gtest_filter_str,
+                                                     filter_pattern_len);
+
+    // Advance gtest_filter_str by filter_pattern_len chars
+    gtest_filter_str += filter_pattern_len;
+  }
+}
+
+// TestNameMatchesGTestFilter(test_name) returns true if any of the following
+// are true:
+// - test_name matches the filter passed in by the last --gtest_filter= command
+//   line argument if any --gtest_filter= arguments are present on the command
+//   line
+// - test_name matches the filter set by the GTEST_FILTER environment variable
+//   if no --gtest_filter= commands were passed into the command line and
+//   the GTEST_FILTER environment variable is set
+// - there were no --gtest_filter= arguments on the command line and the
+//   GTEST_FILTER environment variable is not set
+//
+// Otherwise, TestNameMatchesGTestFilter(test_name) returns false
+static HWY_INLINE HWY_MAYBE_UNUSED bool TestNameMatchesGTestFilter(
+    const char* test_name, size_t test_name_len) {
+  for (const GTestFilterPattern& negative_pattern :
+       NegativeGTestFilterPatterns()) {
+    if (negative_pattern.Matches(test_name, test_name_len)) {
+      return false;
+    }
+  }
+  for (const GTestFilterPattern& positive_pattern :
+       PositiveGTestFilterPatterns()) {
+    if (positive_pattern.Matches(test_name, test_name_len)) {
+      return true;
+    }
+  }
+  return false;
+}
+static HWY_INLINE HWY_MAYBE_UNUSED bool TestNameMatchesGTestFilter(
+    const char* test_name) {
+  return TestNameMatchesGTestFilter(test_name, strlen(test_name));
+}
+static HWY_INLINE HWY_MAYBE_UNUSED bool TestNameMatchesGTestFilter(
+    const std::string& test_name) {
+  return TestNameMatchesGTestFilter(test_name.data(), test_name.length());
+}
+
+// InitTestProgramOptions processes the GTEST_FILTER environment variable, the
+// --gtest_filter= command line argument, and the --gtest_list_tests command
+// line argument
+static HWY_MAYBE_UNUSED void InitTestProgramOptions(
+    const int argc, const char* const* const argv) {
+  // Suppress warning that is normally emitted by MSVC by the getenv call below
+  HWY_DIAGNOSTICS(push)
+#if HWY_COMPILER_MSVC || HWY_COMPILER_CLANGCL
+  HWY_DIAGNOSTICS_OFF(disable : 4996, ignored "-Wdeprecated-declarations")
+#endif
+  const char* gtest_filter = getenv("GTEST_FILTER");
+  HWY_DIAGNOSTICS(pop)
+
+  if (!gtest_filter) {
+    gtest_filter = "*";
+  }
+  for (int i = 1; i < argc; i++) {
+    const char* const curr_arg = argv[i];
+    if (!curr_arg) {
+      break;
+    }
+    if (curr_arg[0] == '-' && curr_arg[1] == '-' && curr_arg[2] == 'g' &&
```

Comment:
What do you think about hoisting these comparisons into a function that takes two const char*? (for matching list_test's \0, we'd have to do this outside the new function because the test string always has \0)
I'd think shorter/more typical code is preferable for arg parsing.